Code/Tools/AssetProcessor/native/utilities/ApplicationManagerBase.cpp

```diff
@@ -449,50 +449,54 @@ void ApplicationManagerBase::InitFileMonitor()
     {
         const auto cachePath = QDir::toNativeSeparators(cacheRoot.absolutePath());
 
+        // For the handlers below, we need to make sure to use invokeMethod on any QObjects so Qt can queue
+        // the callback to run on the QObject's thread if needed.  The APM methods for example are not thread-safe.
+
         const auto OnFileAdded = [this, cachePath](QString path)
         {
             const bool isCacheRoot = path.startsWith(cachePath);
-            if (isCacheRoot)
-            {
-                m_fileStateCache->AddFile(path);
-            }
-            else
+            if (!isCacheRoot)
             {
-                m_assetProcessorManager->AssessAddedFile(path);
-                m_fileStateCache->AddFile(path);
+                bool result = QMetaObject::invokeMethod(m_assetProcessorManager, "AssessAddedFile", Q_ARG(QString, path));
+                AZ_Assert(result, "Failed to invoke m_assetProcessorManager::AssessAddedFile");
+
+                result = QMetaObject::invokeMethod(m_fileProcessor.get(), "AssessAddedFile", Q_ARG(QString, path));
+                AZ_Assert(result, "Failed to invoke m_fileProcessor::AssessAddedFile");
+                
                 AZ::Interface<AssetProcessor::ExcludedFolderCacheInterface>::Get()->FileAdded(path);
-                m_fileProcessor->AssetProcessor::FileProcessor::AssessAddedFile(path);
             }
+
+            m_fileStateCache->AddFile(path);
         };
 
         const auto OnFileModified = [this, cachePath](QString path)
         {
             const bool isCacheRoot = path.startsWith(cachePath);
-            if (isCacheRoot)
-            {
-                m_assetProcessorManager->AssessModifiedFile(path);
-            }
-            else
+            if (!isCacheRoot)
             {
-                m_assetProcessorManager->AssessModifiedFile(path);
                 m_fileStateCache->UpdateFile(path);
             }
+
+            bool result = QMetaObject::invokeMethod(m_assetProcessorManager, "AssessModifiedFile", Q_ARG(QString, path));
+
+            AZ_Assert(result, "Failed to invoke m_assetProcessorManager::AssessModifiedFile");
```

Comment:
Is this a fatal event needing an AZ_Assert, or is it alright to just log this?

Reply:
Same as above, it needs to be fixed by a programmer.  We don't really have any separation between programmer errors and data errors, so there's not really an option besides Assert for programmer errors.  AP would also be pretty useless without these events.